### ISSUE TITLE
Problem parsing floats in cpp

### ISSUE THREAD

**Issue by jgm** [jgm/highlighting-kate#57]
(from @johguenther https://github.com/jgm/pandoc/issues/1879)

Some float constants in code blocks are rendered partly in red when syntax highlighting is used with pandoc 1.13.2, because not all float literals are detected as such. For example, the following markdown

    fails:

    ~~~cpp
    0.1f
    1.0f
    -0.1f
    -1.0F
    -1.0L
    ~~~

    works:

    ~~~cpp
    1e3
    0.f
    1.F
    ~~~

is [translated to the following LaTeX](http://johnmacfarlane.net/pandoc/try/?text=fails%3A%0A%0A~~~cpp%0A0.1f%0A1.0f%0A-0.1f%0A-1.0F%0A-1.0L%0A~~~%0A%0Aworks%3A%0A%0A~~~cpp%0A1e3%0A0.f%0A1.F%0A~~~&from=markdown&to=latex) (where it is easy to see that not all floats are entirely wrapped in `\FloatTok{}`):

    fails:

    \begin{Shaded}
    \begin{Highlighting}[]
    \FloatTok{0.}\ErrorTok{1f}
    \FloatTok{1.}\ErrorTok{0f}
    \FloatTok{-0.}\ErrorTok{1f}
    \FloatTok{-1.}\ErrorTok{0F}
    \FloatTok{-1.}\DecValTok{0L}
    \end{Highlighting}
    \end{Shaded}

    works:

    \begin{Shaded}
    \begin{Highlighting}[]
    \FloatTok{1e3}
    \FloatTok{0.f}
    \FloatTok{1.F}
    \end{Highlighting}
    \end{Shaded}

Expected output is

    \begin{Highlighting}[]
    \FloatTok{0.1f}
    \FloatTok{1.0f}
    \FloatTok{-0.1f}
    \FloatTok{-1.0F}
    \FloatTok{-1.0L}
    \end{Highlighting}

**Comment by jgm** [comment]
The kate highlighting definition for CPP contains

```
        <RegExpr attribute="Float" context="#stay" String="[\+\-]?([0-9]+[Ee][\+\-]?[0-9]+|([0-9]+\.|\.[0-9]+|[0-9]+\.[0-9]+)([Ee][\+\-]?[0-9]+)?)[FfLl]?"
        />
```

This is converted to what should be an equivalent regex in the generated Haskell module:

```
compileRegex True
"[\\+\\-]?([0-9]+[Ee][\\+\\-]?[0-9]+|([0-9]+\\.|\\.[0-9]+|[0-9]+\\.[0-9]+)([Ee][\\+\\-]?[0-9]+)?)[FfLl]?"
```

As far as I can see, this looks like it should match your floats.  So further investigation will be needed.

**Comment by johguenther** [comment]
I think I found the problem with the regex.
The `([0-9]+\.|\.[0-9]+|[0-9]+\.[0-9]+)` part is supposed to match a [nonempty sequence of decimal digits containing a decimal point character](http://en.cppreference.com/w/cpp/language/floating_literal). However, if multiple branches (separated by `|`) match, the first one is used. Thus, the more general branch should come first, i.e. when digits are on both sides of the decimal point.
The corrected regex should be `[\+\-]?([0-9]+[Ee][\+\-]?[0-9]+|([0-9]+\.[0-9]+|[0-9]+\.|\.[0-9]+)([Ee][\+\-]?[0-9]+)?)[FfLl]?`.

**Comment by johguenther** [comment]
Maybe even simpler (and symmetric): `[\+\-]?([0-9]+[Ee][\+\-]?[0-9]+|([0-9]*\.[0-9]+|[0-9]+\.[0-9]*)([Ee][\+\-]?[0-9]+)?)[FfLl]?`

**Comment by jgm** [comment]
Great!  I'm really curious whether the highlighting works properly in
the Kate editor, since the regex really does seem wrong.  (It could be
though that there are differences in the regex engines.)

+++ Johannes Günther [Jan 14 15 09:50 ]:

**Comment by jgm** [comment]
Strangely, it works in Kate.  It must be a difference in how the regex engines work (first match vs longest match).  So we'll have to do a local patch.

**Comment by jgm** [comment]
If you're able to install from source, could you test this?  It seems to work well on the examples you gave.

If you can't easily compile from source (see README for instructions), I can push a new release.

**Comment by johguenther** [comment]
I compiled pandoc with the git-version of highlighting-kate and it worked fine now. Thanks!

Note that I needed to patch the Makefile of highlighting-kate to make `make prep` work:
```diff
diff --git a/Makefile b/Makefile
index 75d454a..3764c71 100644
--- a/Makefile
+++ b/Makefile
@@ -30,7 +30,7 @@ test:
        cabal test
 
 ParseSyntaxFiles: ParseSyntaxFiles.hs
-       cabal install HXT
+       cabal install HXT regex-posix
        ghc --make -Wall ParseSyntaxFiles.hs  # requires HXT >= 9.0.0
 
 clean:
```

**Comment by jgm** [comment]
+++ Johannes Günther [Jan 23 15 05:52 ]:
>I compiled pandoc with the git-version of highlighting-kate and it worked fine now. Thanks!
>
>Note that I needed to patch the Makefile of highlighting-kate to make `make prep` work:

Why?  We shouldn't be using regex-posix anywhere.

**Comment by johguenther** [comment]
Well, before I got
~~~bash
...
ghc --make -Wall ParseSyntaxFiles.hs  # requires HXT >= 9.0.0

ParseSyntaxFiles.hs:36:8:
    Could not find module ‘Text.Regex.Posix’
~~~

**Comment by jgm** [comment]
Ah yes, ParseSyntaxFiles does use it.

**Comment by jgm** [comment]
I added the Makefile patch.

**Comment by johguenther** [comment]
Unfortunately the fix for float parsing got reverted in bcb787e5a1c623cb1f593401bffa9daa9a8d4332. Please re-apply [`isocpp.xml.patch`](../blob/master/xml/isocpp.xml.patch).

**Comment by johguenther** [comment]
Unfortunately the issue surfaced again in the latest pandoc.

**Comment by jgm** [comment]
They moved the regex out of isocpp.xml and are now just using the `<Float>` element.
Unfortunately, skylighting (the successor of highlighting-kate) uses the flawed original regex for this.  I will change it.

**Comment by jgm** [comment]
Should be fixed in skylighting 0.3.1, if you want to build pandoc against that.
